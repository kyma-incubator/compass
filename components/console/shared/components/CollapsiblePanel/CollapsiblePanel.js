import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Panel, Button } from 'fundamental-react';
import './CollapsiblePanel.scss';

export const CollapsiblePanel = ({
  children,
  title,
  className,
  actions,
  isOpenInitially = true,
}) => {
  const [isOpen, setIsOpen] = useState(isOpenInitially);

  const switchOpen = e => {
    e.stopPropagation();
    // ensure event didn't come from DOM propagation
    if (e.target === e.currentTarget) {
      setIsOpen(!isOpen);
    }
  };

  return (
    <Panel className={classNames('collapsible-panel', className)}>
      <Panel.Header onClick={switchOpen}>
        <h3 className="fd-panel__title" onClick={switchOpen}>
          {title}
        </h3>
        <Panel.Actions>
          {actions}
          <Button
            glyph={isOpen ? 'navigation-up-arrow' : 'navigation-down-arrow'}
            option="light"
            onClick={switchOpen}
          />
        </Panel.Actions>
      </Panel.Header>
      <Panel.Body className={isOpen ? 'body body--open' : 'body body--closed'}>
        {children}
      </Panel.Body>
    </Panel>
  );
};

CollapsiblePanel.propTypes = {
  children: PropTypes.element.isRequired,
  title: PropTypes.string.isRequired,
  className: PropTypes.string,
  actions: PropTypes.element,
  isOpenInitially: PropTypes.bool,
};
