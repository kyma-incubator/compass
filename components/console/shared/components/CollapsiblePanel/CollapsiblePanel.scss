.collapsible-panel {
    .body{
        transition: height 1s, padding 0.5s, opacity 1s ease-in, transform 0.3s ease-out;

        &--open {
            max-height:1000em;
        }

        &--closed {
            transform: scaleY(0) translateY(-100%);
            max-height:0px;
            padding-top:0;
            padding-bottom:0;
            overflow: hidden;
        }
    }

    .fd-panel__header {
        cursor: pointer;
    }
}