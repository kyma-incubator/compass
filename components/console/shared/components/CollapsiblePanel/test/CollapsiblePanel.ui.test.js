import React from 'react';
import renderer from 'react-test-renderer';
import { Button, Panel } from 'fundamental-react';
import { mount, shallow } from 'enzyme';

import { CollapsiblePanel } from './../CollapsiblePanel';

describe('Collapsible Panel', () => {
  const exceptPanelVisible = (component, isVisible) => {
    const body = component.find(Panel.Body);
    expect(body.hasClass('body--closed')).toBe(!isVisible);
    expect(body.hasClass('body--open')).toBe(isVisible);
  };

  it('Renders with minimal props (opened state)', () => {
    const component = renderer.create(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel test"
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders in closed state', () => {
    const component = renderer.create(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel test"
        isOpenInitially={false}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Applies custom class names to Panel', () => {
    const component = shallow(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel"
        className="test-class-name"
      />,
    );
    expect(component.find(Panel).hasClass('test-class-name')).toBe(true);
  });

  it('Is initially closed when isOpenInitially is false', () => {
    const component = shallow(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel"
        isOpenInitially={false}
      />,
    );

    expect(component.find(Panel.Body).hasClass('body body--closed')).toBe(true);
    expect(component.find(Panel.Body).hasClass('body body--open')).toBe(false);
  });

  it('Opens and closes (chevron)', async () => {
    const component = mount(
      <CollapsiblePanel children={<p>test</p>} title="Collapsible panel" />,
    );
    const button = component.find(Button);

    // initially opened
    exceptPanelVisible(component, true);

    // close
    button.simulate('click');
    await component.update();
    exceptPanelVisible(component, false);

    // open again
    button.simulate('click');
    await component.update();
    exceptPanelVisible(component, true);
  });

  it('Opens and closes (header)', async () => {
    const component = mount(
      <CollapsiblePanel children={<p>test</p>} title="Collapsible panel" />,
    );
    const header = component.find(Panel.Header);

    // initially opened
    exceptPanelVisible(component, true);

    // close
    header.simulate('click');
    await component.update();
    exceptPanelVisible(component, false);

    // open again
    header.simulate('click');
    await component.update();
    exceptPanelVisible(component, true);
  });

  it('Does not open modal when clicking on custom actions', async () => {
    const child = <p>test</p>;
    const component = mount(
      <CollapsiblePanel
        children={child}
        title="Collapsible panel"
        actions={
          <button id="action-button" onClick={() => {}}>
            test
          </button>
        }
      />,
    );

    const button = component.find('#action-button');

    // close
    button.simulate('click');
    await component.update();

    // expect it's still opened
    exceptPanelVisible(component, true);
  });
});
