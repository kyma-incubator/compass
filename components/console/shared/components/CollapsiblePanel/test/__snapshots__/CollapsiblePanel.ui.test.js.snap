// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible Panel Renders in closed state 1`] = `
<div
  className="fd-panel collapsible-panel"
>
  <div
    className="fd-panel__header"
    onClick={[Function]}
  >
    <h3
      className="fd-panel__title"
      onClick={[Function]}
    >
      Collapsible panel test
    </h3>
    <div
      className="fd-panel__actions"
    >
      <button
        className="fd-button--light sap-icon--navigation-down-arrow"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="fd-panel__body body body--closed"
  >
    <p>
      test
    </p>
  </div>
</div>
`;

exports[`Collapsible Panel Renders with minimal props (opened state) 1`] = `
<div
  className="fd-panel collapsible-panel"
>
  <div
    className="fd-panel__header"
    onClick={[Function]}
  >
    <h3
      className="fd-panel__title"
      onClick={[Function]}
    >
      Collapsible panel test
    </h3>
    <div
      className="fd-panel__actions"
    >
      <button
        className="fd-button--light sap-icon--navigation-up-arrow"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="fd-panel__body body body--open"
  >
    <p>
      test
    </p>
  </div>
</div>
`;
